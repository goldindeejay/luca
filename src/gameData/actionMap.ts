import { BattleData } from './battleData';

import * as _ from 'lodash';

/**
 * An actionMap item from battle.js's scenes/battle/AbilityFactory module.
 */
export interface ActionMapItem {
  actionId: number;
  className: string;
  changeCastTimeCondList?: string[];
  elements?: {
    args: number[];
  };
  isAttack?: boolean;
  isHeal?: boolean;
  isHealHp?: boolean;
  isTrance?: boolean;
  enemyTargeting?: number[];
  setSa?: {
    useStatusAilmentsId?: boolean;
    autoConvertSaIdToBundle?: boolean;
    args?: number[];
    bundleArgs: number[];
  };
  unsetSa?: {
    useStatusAilmentsId?: boolean;
    autoConvertSaIdToBundle?: boolean;
    args?: number[];
    bundleArgs: number[];
  };
  isBraveMode?: boolean;
  isPossibleContainMagicDamage?: boolean;
  ignoresMirageAndMightyGuardArg?: number;
  ignoresReflectionArg?: number;
  ignoresStatusAilmentsBarrierArg?: number;
  burstAbilityArgs?: number[];
  isFlightAttack?: boolean;
}

export interface ActionLookup {
  [i: number]: ActionMapItem;
}

export function makeActionLookup(battleData: BattleData): { [i: number]: ActionMapItem } {
  return _.fromPairs(
    battleData.ns.battle.AbilityFactory.actionMap.map((i: any) => [i.actionId, i]),
  ) as any;
}
