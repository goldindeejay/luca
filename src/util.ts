export const toEuroFixed = (value: number) => value.toFixed(2).replace('.', ',');
export const withPlus = (value: number) => (value < 0 ? value : '+' + value);
