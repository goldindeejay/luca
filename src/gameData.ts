import * as _ from 'lodash';
import { vsprintf } from 'sprintf-js';

import { NamedArgs } from './battleActionDetails';
import { LangType } from './util';

/**
 * Limited definition of battle.json, which we process from battle.js then add
 * our own `extra` property.  It's involved enough that we don't want to
 * duplicate all of that in TypeScript right now.
 */
interface BattleDataType {
  ns: {
    battle: {
      util: any;
      Conf: any;
      AbilityFactory: any;
      Config: any;
      StatusAilmentConfig: any;
    };
  };
  env: {};
  extra: {
    statusAilments: any;
    statusAilmentBundles: any;
  };
}

const rawBattleData: { [lang in LangType]: BattleDataType } = {
  [LangType.Gl]: require('./gl/battle.json'),
  [LangType.Jp]: require('./jp/battle.json'),
};

function makeLookup<T extends string>(enumType: any) {
  const result: { [s: string]: T } = {};
  for (const i of Object.keys(enumType)) {
    result[enumType[i as any]] = i as T;
  }
  return result;
}

/**
 * Damage codes.  Aka "EXERCISE_TYPE."  These can be found in battle.js.
 */
export enum DamageType {
  // noinspection JSUnusedGlobalSymbols
  PHY = 1,
  WHT = 3,
  BLK = 4,
  BLU = 5,
  SUM = 6,
  NAT = 7, // aka "Inborn"
  NIN = 8,
  NONE = 9,
}

export enum ElementType {
  // noinspection JSUnusedGlobalSymbols
  Fire = 100,
  Ice = 101,
  Lightning = 102,
  Earth = 103,
  Wind = 104,
  Water = 105,
  Holy = 106,
  Dark = 107,
  Poison = 108,
  NE = 199,
}

const schoolAlias: { [key: string]: string } = {
  SHOOTER: 'Sharpshooter',
};

/**
 * Maps target_range + target_segment values to [full description, noun]
 * tuples.
 */
const targetLookup: { [range: string]: { [segment: string]: [string, string] } } = {
  SINGLE: {
    OPPONENT: ['Single enemy', 'enemy'],
    COLLEAGUE: ['Single ally', 'ally'],
    BOTH: ['Single', ''],
    BOTH_EXCEPT_MYSELF: ['Another', 'other'],
    COLLEAGUE_EXCEPT_MYSELF: ['Another ally', 'other ally'],
  },
  ALL: {
    OPPONENT: ['All enemies', 'enemies'],
    COLLEAGUE: ['All allies', 'allies'],
    BOTH: ['All', ''],
    BOTH_EXCEPT_MYSELF: ['All others', 'others'],
    COLLEAGUE_EXCEPT_MYSELF: ['All other allies', 'other allies'],
  },
};

// noinspection SpellCheckingInspection
const targetMethodDescription: { [targetMethod: string]: string | null } = {
  HP_RATIO_DESC: 'Highest HP%% %s',
  HP_RATIO_ASC: 'Lowest HP%% %s',
  SA_RANDOM: 'Random %s without status',
  DIS_SA_RANDOM: 'Random %s without status',
  RANDOM: 'Random %s',
  NOTHING: null,
  HP_DESC: 'Lowest HP %s',
  HP_ASC: 'Highest HP %s',
  ESNA: 'Random %s with status',
  DISPEL: 'Random %s with status',

  // Unconfirmed
  MP_RANDOM: '%s',

  // Used by, e.g., Porom's Curaise BSB CMD1
  MARAISE: '%1$s with KO or lowest HP%% %1$s',

  // Unconfirmed
  LOT_BY_HP_RATE: 'Highest HP%% %s',

  // Used by, e.g., Death.  It perhaps means "vulnerable enemy"?
  SA_DEF_ATTRIBUTE_COUNT_AND_HAS_SA_BUNDLE_COUNT_SUM_ASC: 'Random %s',

  // Unconfirmed - treating as above for now
  SA_PARALYSIS_AND_SILENCE_DEF_ATTRIBUTE_COUNT_AND_HAS_SA_BUNDLE_COUNT_SUM_ASC: 'Random %s',

  // These are all unconfirmed.
  BUDDY_SMART: 'Smart %s',
  AI_SMART: 'Smart %s',
  AI_SMART_IGNORE_REFRECTION: 'Smart %s',
};

function makeBattleDataHelpers(lang: LangType) {
  const data = rawBattleData[lang];
  const conf = data.ns.battle.Conf;

  const result = {
    conf,
    damageTypeLookup: makeLookup(DamageType),
    elementTypeLookup: makeLookup(ElementType),
    attackId: conf.ABILITY_ID_OF.ATTACK,
    isAprilFoolId: (id: number) => id === conf.ABILITY_ID_OF.SKETCH_FOR_APRIL_FOOL_2017,
    schoolTypeLookup: _.fromPairs(
      _.map(conf.ABILITY_CATEGORY_ID, (value, key) => [
        value,
        schoolAlias[key] || _.startCase(_.toLower(key)),
      ]),
    ),
    targetRangeLookup: _.invert(conf.TARGET_RANGE),
    targetSegmentLookup: _.invert(conf.TARGET_SEGMENT),
    activeTargetMethodLookup: _.invert(conf.ACTIVE_TARGET_METHOD),
    targetMethodLookup: _.invert(conf.TARGET_METHOD),

    describeTarget(
      rangeValue: string | number,
      segmentValue: string | number,
      activeTargetMethodValue: string | number | undefined,
    ): string | null {
      const range = result.targetRangeLookup[rangeValue];
      const segment = result.targetSegmentLookup[segmentValue];
      const activeMethod = activeTargetMethodValue
        ? result.activeTargetMethodLookup[activeTargetMethodValue]
        : null;
      // This function is ugly - I'm still unclear on the relationship between
      // target segment and active target method.  For example, Ultra Cure gives
      // a segment of colleague, but its active target method allows it to
      // target enemies.
      if (range === 'SELF') {
        return 'Self';
      } else if (range === 'SINGLE' && activeMethod && activeMethod !== 'BOTH_DISABLE') {
        return activeMethod === 'BOTH_ENABLE'
          ? 'Single'
          : activeMethod === 'OPPONENT_DISABLE'
          ? 'Single ally'
          : 'Single enemy';
      } else if (range === 'SINGLE' && segment === 'OPPONENT' && activeMethod === 'BOTH_DISABLE') {
        return 'Random enemies';
      } else if (targetLookup[range]) {
        const lookupResult = targetLookup[range][segment];
        return lookupResult ? lookupResult[0] : null;
      } else {
        return null;
      }
    },

    describeTargetMethod(
      rangeValue: string | number,
      segmentValue: string | number,
      activeTargetMethodValue: string | number | undefined,
      targetMethodValue: string | number,
    ): string | null {
      const range = result.targetRangeLookup[rangeValue];
      const segment = result.targetSegmentLookup[segmentValue];
      const method = result.targetMethodLookup[targetMethodValue];
      const methodDescription = targetMethodDescription[method];
      if (range === 'SELF' || !methodDescription) {
        return result.describeTarget(rangeValue, segmentValue, activeTargetMethodValue);
      } else {
        const lookupResult = targetLookup[range][segment];
        return lookupResult
          ? _.upperFirst(vsprintf(methodDescription, [lookupResult[1]])).trim()
          : null;
      }
    },
  };

  return result;
}

/**
 * Battle action details as automatically processed from battle.js by
 * get-battle-args.js.  Properly belongs in battleActionDetails.js, but putting
 * it here lets us keep all of our JSON data grouped together.
 */
export interface BattleActionArgs {
  /**
   * Map from args1..args30 to named arguments.
   */
  args: { [key in keyof NamedArgs]: number };

  multiArgs?: { [key in keyof NamedArgs]: number[] };
}

// FIXME: This isn't type-checked properly - TypeScript lets me pass a BattleData in place of an ActionMap
type BattleDataHelpers = ReturnType<typeof makeBattleDataHelpers>;
export type BattleData = BattleDataType &
  BattleDataHelpers & {
    battleActionArgs: { [actionName: string]: BattleActionArgs };
  };

export const battleData: { [lang in LangType]: BattleData } = {
  [LangType.Gl]: {
    ...rawBattleData[LangType.Gl],
    ...makeBattleDataHelpers(LangType.Gl),
    battleActionArgs: require('./gl/battleArgs.json'),
  },
  [LangType.Jp]: {
    ...rawBattleData[LangType.Jp],
    ...makeBattleDataHelpers(LangType.Jp),
    battleActionArgs: require('./jp/battleArgs.json'),
  },
};
