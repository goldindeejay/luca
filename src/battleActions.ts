import { BattleActionDetails, battleActionDetails, NamedArgs } from './battleActionDetails';
import { BattleActionArgs, BattleData } from './gameData';
import { logger } from './logger';
import {
  BeastActiveSkill,
  BuddyAbility,
  BuddySoulStrike,
  Options,
} from './schemas/get_battle_init_data';

import * as _ from 'lodash';

interface ActionMapItem {
  actionId: number;
  className: string;
  changeCastTimeCondList?: string[];
  elements?: {
    args: number[];
  };
  isAttack?: boolean;
  isHeal?: boolean;
  isHealHp?: boolean;
  isTrance?: boolean;
  enemyTargeting?: number[];
  setSa?: {
    useStatusAilmentsId?: boolean;
    autoConvertSaIdToBundle?: boolean;
    args?: number[];
    bundleArgs: number[];
  };
  unsetSa?: {
    useStatusAilmentsId?: boolean;
    autoConvertSaIdToBundle?: boolean;
    args?: number[];
    bundleArgs: number[];
  };
  isBraveMode?: boolean;
  isPossibleContainMagicDamage?: boolean;
  ignoresMirageAndMightyGuardArg?: number;
  ignoresReflectionArg?: number;
  ignoresStatusAilmentsBarrierArg?: number;
  burstAbilityArgs?: number[];
}

interface ActionLookup {
  [i: number]: ActionMapItem;
}

export function makeActionLookup(battleData: BattleData): { [i: number]: ActionMapItem } {
  return _.fromPairs(
    battleData.ns.battle.AbilityFactory.actionMap.map((i: any) => [i.actionId, i]),
  ) as any;
}

function logMissing(actionLookup: ActionLookup, actionId: number) {
  if (!actionLookup[actionId]) {
    logger.warn(`Unknown action ID ${actionId}`);
  } else {
    logger.warn(`Missing details for action ID ${actionId} (${actionLookup[actionId].className})`);
  }
}

function getArgs(options: Options): number[] {
  const result = [];
  for (let i = 1; i <= 30; i++) {
    const arg = `arg${i}`;
    if (options[arg] == null) {
      throw new Error(`Missing ${arg}`);
    }

    const value = options[arg];
    if (typeof value !== 'string' || isNaN(Number(value)) || !Number.isInteger(+value)) {
      throw new Error(`Bad ${arg} "${value}`);
    }

    result[i] = +value;
  }
  return result;
}

/**
 * Gets details about battle actions for the given action ID.
 *
 * Returns the following:
 * - The action name, if known
 * - The action argument mappings, as automatically extracted from battle.js
 * - The manually maintained action details, including argument mappings,
 *   formatters, etc.
 */
export function getBattleActionDetails(
  battleData: BattleData,
  actionLookup: ActionLookup,
  actionId: number,
): [ActionMapItem | null, BattleActionArgs | null, BattleActionDetails | null] {
  const action = actionLookup[actionId];
  if (!action) {
    return [null, null, null];
  }

  const actionName = action.className;
  const args = battleData.battleActionArgs[actionName];
  const details = battleActionDetails[actionName];
  return [action, args || null, details || null];
}

/**
 * Maps from an FFRK options object (with values like arg1, arg2...) to our
 * NamedArgs (which contains the same values, but with meaningful property
 * names, courtesy of battle.js actions).
 */
export function getNamedArgs(
  battleData: BattleData,
  actionLookup: ActionLookup,
  actionId: number,
  options: Options,
): NamedArgs | null {
  const [action, actionArgs, details] = getBattleActionDetails(battleData, actionLookup, actionId);
  if (!action || !actionArgs || !details) {
    logMissing(actionLookup, actionId);
  }
  if (!action || !actionArgs) {
    return null;
  }
  const argSource = details || actionArgs;

  const args = getArgs(options);

  // Map arguments that are specified in FFRK JS code and that we manually
  // define in our own TS code to match.
  // FIXME: This is ugly.  Refactor and unit test.
  const result: NamedArgs = {
    ..._.fromPairs(_.map(argSource.args, (value, key) => [key, args[value as number]])),
    ..._.fromPairs(
      _.map(argSource.multiArgs, (value, key) => [
        key,
        _.filter((value as number[]).map((i: number) => args[i])),
      ]),
    ),
  };

  // Map arguments that are specified in FFRK actionMap options.
  function tryArg(key: keyof NamedArgs, argNumber: number | undefined) {
    if (argNumber) {
      result[key] = args[argNumber];
    }
  }
  function tryArgList(key: keyof NamedArgs, argNumbers: number[] | undefined) {
    if (argNumbers && argNumbers.length) {
      result[key] = _.filter(argNumbers.map(i => args[i]));
    }
  }
  tryArgList('elements', action.elements && action.elements.args);
  tryArg('ignoresReflection', action.ignoresReflectionArg);
  tryArg('ignoresMirageAndMightyGuard', action.ignoresMirageAndMightyGuardArg);
  tryArg('ignoresStatusAilmentsBarrier', action.ignoresStatusAilmentsBarrierArg);
  tryArgList('burstAbility', action.burstAbilityArgs);
  tryArgList('setSaId', action.setSa && action.setSa.args);
  tryArgList('setSaBundle', action.setSa && action.setSa.bundleArgs);
  tryArgList('unsetSaId', action.unsetSa && action.unsetSa.args);
  tryArgList('unsetSaBundle', action.unsetSa && action.unsetSa.bundleArgs);

  return result;
}

export function getMultiplier(args: NamedArgs | null): number | null {
  if (args && args.damageFactor) {
    return ((args.barrageNum || 1) * args.damageFactor) / 100;
  } else {
    return null;
  }
}

export function getElements(battleData: BattleData, args: NamedArgs | null): string | null {
  if (!args) {
    return null;
  } else if (args.elements && args.elements.length) {
    return args.elements.map(i => battleData.elementTypeLookup[i]).join(', ');
  } else if (args.matkElement) {
    return battleData.elementTypeLookup[args.matkElement];
  } else {
    return null;
  }
}

export function getAbilityDescription(
  battleData: BattleData,
  actionLookup: ActionLookup,
  actionId: number,
  options: Options,
  args: NamedArgs | null,
): string | null {
  const [, , details] = getBattleActionDetails(battleData, actionLookup, actionId);
  if (details && args) {
    return details.formatEnlir(battleData, options, args);
  } else {
    return null;
  }
}

const toBool = (value: string | number) => !!+value;
const toBoolOrNull = (value: string | number | undefined) => (value == null ? null : toBool(value));
const msecToSec = (msec: string | number) => +msec / 1000;

/**
 * High-level function to convert JSON for a game "ability" (including soul
 * break or magicite skill) to JSON.
 */
export function convertAbility(
  battleData: BattleData,
  abilityData: BuddyAbility | BuddySoulStrike | BeastActiveSkill,
): any {
  const { options } = abilityData;
  const actionLookup = makeActionLookup(battleData);

  let alias = null;
  if (options.alias_name !== '' && options.alias_name !== options.name) {
    alias = options.alias_name;
  }

  if (+abilityData.ability_id === battleData.attackId && options.name === 'Attack') {
    return null;
  }
  if (battleData.isAprilFoolId(+abilityData.ability_id)) {
    return null;
  }

  const school = abilityData.category_id
    ? battleData.schoolTypeLookup[+abilityData.category_id] || null
    : null;
  const [action, , details] = getBattleActionDetails(
    battleData,
    actionLookup,
    +abilityData.action_id,
  );
  const args = getNamedArgs(battleData, actionLookup, +abilityData.action_id, options);

  // Not yet used:
  // const breaksDamageCap = toBool(options.max_damage_threshold_type);

  // Omit options.target_death; it corresponds to TARGET_DEATH, but abilities'
  // effects make it obvious whether they can target dead allies.

  // Not included: rarity, uses, max, orbs
  // Those apply to what the game UI calls abilities, but what the game UI
  // calls abilities is not the same as what the game JSON calls abilities.

  // Also not included: nameJp, gl, introducingEvent
  // Those need to be handled at a higher level and likely require outside
  // data and/or human intervention.
  return {
    school,
    name: options.name.trim(),
    alias: alias ? alias.trim() : alias,
    type: battleData.damageTypeLookup[+abilityData.exercise_type],
    target: battleData.describeTarget(
      options.target_range,
      options.target_segment,
      options.active_target_method,
    ),
    formula: details ? details.formula : null,
    multiplier: getMultiplier(args),
    element: getElements(battleData, args),
    time: msecToSec(options.cast_time),
    effects: getAbilityDescription(battleData, actionLookup, +abilityData.action_id, options, args),
    counter: toBoolOrNull(options.counter_enable),
    autoTarget: battleData.describeTargetMethod(
      options.target_range,
      options.target_segment,
      options.active_target_method,
      options.target_method,
    ),
    sb: options.ss_point == null ? null : +options.ss_point,
    id: +abilityData.ability_id,
    action: action ? action.className : null,
    args: getNamedArgs(battleData, actionLookup, +abilityData.action_id, options),
  };
}
