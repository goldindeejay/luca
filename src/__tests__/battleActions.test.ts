import { convertAbility } from '../battleActions';
import { battleData } from '../gameData/battleData';
import { LangType } from '../util';

describe('battleActions', () => {
  describe('convertAbility', () => {
    it('converts simple character abilities', () => {
      const dreadWeaponAbilityData = {
        category_id: '17',
        options: {
          ability_animation_id: '10267',
          active_target_method: '3',
          alias_name: '',
          arg1: '75',
          arg10: '0',
          arg11: '0',
          arg12: '0',
          arg13: '0',
          arg14: '0',
          arg15: '0',
          arg16: '0',
          arg17: '0',
          arg18: '0',
          arg19: '0',
          arg2: '4',
          arg20: '0',
          arg21: '0',
          arg22: '0',
          arg23: '0',
          arg24: '0',
          arg25: '0',
          arg26: '0',
          arg27: '0',
          arg28: '0',
          arg29: '0',
          arg3: '1',
          arg30: '0',
          arg4: '0',
          arg5: '107',
          arg6: '1',
          arg7: '20',
          arg8: '0',
          arg9: '0',
          cast_time: '1650',
          counter_enable: '1',
          max_damage_threshold_type: '0',
          min_damage_threshold_type: '0',
          name: 'Dread Weapon',
          panel_name: 'Dread{n}Weapon',
          ss_point: '75',
          status_ailments_factor: '0',
          status_ailments_id: '0',
          target_death: '1',
          target_method: '2',
          target_range: '1',
          target_segment: '1',
        },
        exercise_type: '1',
        action_id: '126',
        ability_id: '30271101',
      };
      const ability = convertAbility(battleData[LangType.Gl], dreadWeaponAbilityData);
      expect(ability).toMatchSnapshot();
    });

    it('converts soul break jump attacks', () => {
      const kainCsbEffectAbilityData = {
        category_id: '5',
        options: {
          arg4: '1',
          min_damage_threshold_type: '0',
          arg7: '0',
          arg20: '0',
          status_ailments_id: '0',
          arg18: '0',
          ss_point: '0',
          arg28: '0',
          arg12: '0',
          arg22: '0',
          arg17: '0',
          arg15: '0',
          arg1: '36',
          arg10: '0',
          arg21: '0',
          arg11: '0',
          name: 'Impulse Drive',
          arg24: '0',
          arg2: '22',
          arg5: '102',
          alias_name: 'Impulse Drive',
          panel_name: 'Impulse Drive',
          arg14: '0',
          target_segment: '1',
          arg25: '0',
          arg30: '0',
          arg9: '0',
          arg13: '0',
          arg19: '0',
          arg27: '0',
          counter_enable: '0',
          arg23: '0',
          arg29: '0',
          target_range: '1',
          arg26: '0',
          target_death: '1',
          arg3: '2',
          cast_time: '2500',
          arg8: '0',
          max_damage_threshold_type: '0',
          arg6: '1',
          target_method: '2',
          active_target_method: '3',
          status_ailments_factor: '100',
          arg16: '0',
          ability_animation_id: '71736',
        },
        exercise_type: '1',
        action_id: '7',
        ability_id: '30544011',
      };

      const ability = convertAbility(battleData[LangType.Jp], kainCsbEffectAbilityData);

      expect(ability.effects).toEqual('Twenty-two single ranged jump attacks (0,36 each)');

      expect(ability).toMatchSnapshot();
    });
  });
});
