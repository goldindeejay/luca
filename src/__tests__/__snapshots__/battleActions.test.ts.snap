// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`battleActions convertAbility converts PhysicalAttackMultiAndSelfSaAction abilities 1`] = `
Object {
  "action": "PhysicalAttackMultiAndSelfSaAction",
  "alias": null,
  "args": Object {
    "atkType": 2,
    "barrageNum": 3,
    "critical": 0,
    "damageCalculateParamAdjust": 0,
    "damageCalculateParamAdjustConf": Array [],
    "damageFactor": 70,
    "elements": Array [
      102,
      104,
    ],
    "forceHit": 0,
    "ignoresAttackHit": 1,
    "isSameTarget": 1,
    "selfSaAnimationFlag": 1,
    "selfSaId": 257,
    "selfSaOptionsDuration": 0,
  },
  "autoTarget": "Lowest HP% enemy",
  "breaksDamageCap": false,
  "counter": true,
  "effects": "Three single ranged attacks (0,70 each), grants No Air Time 2 to the user",
  "element": "Lightning, Wind",
  "formula": "Physical",
  "id": 30181131,
  "multiplier": 2.1,
  "name": "Cyclone Bolt",
  "nameGl": null,
  "sb": 75,
  "school": "Dragoon",
  "statusAilmentsFactor": 0,
  "statusAilmentsId": 0,
  "target": "Single enemy",
  "time": 1.5,
  "type": "PHY",
}
`;

exports[`battleActions convertAbility converts abilities that remove status ailments 1`] = `
Object {
  "action": "HealHpAndHealSaAction",
  "alias": null,
  "args": Object {
    "damageFactor": 400,
    "factor": 105,
    "ignoresReflection": 0,
    "matkElement": 106,
    "unsetSaBundle": Array [
      2,
    ],
    "unsetSaId": Array [],
  },
  "autoTarget": "Lowest HP% ally",
  "breaksDamageCap": false,
  "counter": true,
  "effects": "Restores HP (105), damages undeads, removes negative effects",
  "element": "Holy",
  "formula": "Magical",
  "id": 30121291,
  "multiplier": 4,
  "name": "Ultra Cure",
  "nameGl": null,
  "sb": 85,
  "school": "White Magic",
  "statusAilmentsFactor": 0,
  "statusAilmentsId": 0,
  "target": "Single",
  "time": 1.5,
  "type": "WHT",
}
`;

exports[`battleActions convertAbility converts simple character abilities 1`] = `
Object {
  "action": "PhysicalAttackMultiAndHealHpByHitDamageAction",
  "alias": null,
  "args": Object {
    "atkType": 1,
    "barrageNum": 4,
    "damageFactor": 75,
    "elements": Array [
      107,
    ],
    "forceHit": 0,
    "healHpFactor": 20,
    "isSameTarget": 1,
  },
  "autoTarget": "Lowest HP% enemy",
  "breaksDamageCap": false,
  "counter": true,
  "effects": "Four single attacks (0,75 each), heals the user for 20% of the damage dealt",
  "element": "Dark",
  "formula": "Physical",
  "id": 30271101,
  "multiplier": 3,
  "name": "Dread Weapon",
  "nameGl": null,
  "sb": 75,
  "school": "Darkness",
  "statusAilmentsFactor": 0,
  "statusAilmentsId": 0,
  "target": "Single enemy",
  "time": 1.65,
  "type": "PHY",
}
`;

exports[`battleActions convertAbility converts soul break jump attacks 1`] = `
Object {
  "action": "PhysicalAttackMultiAction",
  "alias": null,
  "args": Object {
    "atkElement": 102,
    "atkExponentialFactor": 0,
    "atkType": 2,
    "barrageNum": 22,
    "critical": 0,
    "damageCalculateParamAdjust": 0,
    "damageCalculateParamAdjustConf": Array [],
    "damageCalculateTypeByAbility": 0,
    "damageFactor": 36,
    "forceHit": 1,
    "ignoresMirageAndMightyGuard": 0,
    "isFlightAttack": true,
    "isSameTarget": 1,
    "situationalRecalculateDamageHookType": 0,
  },
  "autoTarget": "Lowest HP% enemy",
  "breaksDamageCap": false,
  "counter": false,
  "effects": "Twenty-two single ranged jump attacks (0,36 each)",
  "element": "Lightning",
  "formula": undefined,
  "id": 30544011,
  "multiplier": 7.92,
  "name": "Impulse Drive",
  "nameGl": null,
  "sb": 0,
  "school": "Combat",
  "statusAilmentsFactor": 100,
  "statusAilmentsId": 0,
  "target": "Single enemy",
  "time": 2.5,
  "type": "PHY",
}
`;
